mithril-aggregator/src/services/aggregator_client.rs

```diff
@@ -0,0 +1,518 @@
+                )
+            } else if let Ok(server_error) =
+                serde_json::from_value::<ServerError>(json_value.clone())
+            {
+                format!("{}: {}", canonical_reason, server_error.message)
+            } else if json_value.is_null() {
+                canonical_reason.to_string()
+            } else {
+                format!("{}: {}", canonical_reason, json_value)
+            }
+        } else {
+            let response_text = response.text().await.unwrap_or_default();
+            format!("{}: {}", canonical_reason, response_text)
+        }
+    }
+}
+
+/// Trait for mocking and testing a `AggregatorClient`
+#[cfg_attr(test, mockall::automock)]
+#[async_trait]
+pub trait AggregatorClient: Sync + Send {
+    /// Retrieves epoch settings from the aggregator
+    async fn retrieve_epoch_settings(
+        &self,
+    ) -> Result<Option<MasterAggregatorEpochSettings>, AggregatorClientError>;
+}
+
+/// AggregatorHTTPClient is a http client for an aggregator
+pub struct AggregatorHTTPClient {
+    aggregator_endpoint: String,
+    relay_endpoint: Option<String>,
+    api_version_provider: Arc<APIVersionProvider>,
+    timeout_duration: Option<Duration>,
+    logger: Logger,
+}
+
+impl AggregatorHTTPClient {
+    /// AggregatorHTTPClient factory
+    pub fn new(
+        aggregator_endpoint: String,
+        relay_endpoint: Option<String>,
+        api_version_provider: Arc<APIVersionProvider>,
+        timeout_duration: Option<Duration>,
+        logger: Logger,
+    ) -> Self {
+        let logger = logger.new_with_component_name::<Self>();
+        debug!(logger, "New AggregatorHTTPClient created");
+        Self {
+            aggregator_endpoint,
+            relay_endpoint,
+            api_version_provider,
+            timeout_duration,
+            logger,
+        }
+    }
+
+    fn prepare_http_client(&self) -> Result<Client, AggregatorClientError> {
+        let client = match &self.relay_endpoint {
+            Some(relay_endpoint) => Client::builder()
+                .proxy(
+                    Proxy::all(relay_endpoint)
+                        .map_err(|e| AggregatorClientError::ProxyCreation(anyhow!(e)))?,
+                )
+                .build()
+                .map_err(|e| AggregatorClientError::HTTPClientCreation(anyhow!(e)))?,
+            None => Client::new(),
+        };
+
+        Ok(client)
+    }
+
+    /// Forge a client request adding protocol version in the headers.
+    pub fn prepare_request_builder(&self, request_builder: RequestBuilder) -> RequestBuilder {
+        let request_builder = request_builder
+            .header(
+                MITHRIL_API_VERSION_HEADER,
+                self.api_version_provider
+                    .compute_current_version()
+                    .unwrap()
+                    .to_string(),
+            )
+            .header(MITHRIL_SIGNER_VERSION_HEADER, env!("CARGO_PKG_VERSION"));
```

Comment:
We should send another header or remove it.

Reply:
Good catch! I used a new `MITHRIL_AGGREGATOR_VERSION_HEADER` header instead :+1: